src/lib/utils/meta_tables/meta_system_utilization_table.cpp

```diff
@@ -0,0 +1,270 @@
+
+  return output_table;
+}
+
+/*
+  * Returns the load average values for 1min, 5min, and 15min.
+*/
+MetaSystemUtilizationTable::LoadAvg MetaSystemUtilizationTable::_get_load_avg() {
+  std::array<double, 3> load_avg{};
+  [[maybe_unused]] const int nelem = getloadavg(load_avg.data(), 3);
+  DebugAssert(nelem == 3, "Failed to read load averages");
+  return {static_cast<float>(load_avg[0]), static_cast<float>(load_avg[1]), static_cast<float>(load_avg[2])};
+}
+
+/*
+  * Returns the time in ns since epoch.
+*/
+uint64_t MetaSystemUtilizationTable::_get_total_time() {
+  auto time = std::chrono::steady_clock::now().time_since_epoch();
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(time).count();
+}
+
+/*
+ * Returns the time in ns that ALL processes have spent on the CPU
+ * since an arbitrary point in the past.
+*/
+uint64_t MetaSystemUtilizationTable::_get_system_cpu_time() {
+#ifdef __linux__
+  std::ifstream stat_file;
+  stat_file.open("/proc/stat", std::ifstream::in);
+  DebugAssert(stat_file.is_open(), "Failed to open /proc/stat");
+
+  std::string cpu_line;
+  std::getline(stat_file, cpu_line);
+  stat_file.close();
+
+  const auto cpu_ticks = _parse_value_string(cpu_line);
+
+  const auto user_ticks = cpu_ticks.at(0);
+  const auto user_nice_ticks = cpu_ticks.at(1);
+  const auto kernel_ticks = cpu_ticks.at(2);
+
+  const auto active_ticks = user_ticks + user_nice_ticks + kernel_ticks;
+  size_t cpu_count;
+  if (numa_available() != -1) {
+    cpu_count = numa_num_task_cpus();
+  } else {
+    cpu_count = _get_cpu_count();
+  }
+
+  // The amount of time in /proc/stat is measured in units of clock ticks.
+  // sysconf(_SC_CLK_TCK) can be used to convert it to ns.
+  const auto active_ns = (active_ticks * std::nano::den) / (sysconf(_SC_CLK_TCK) * cpu_count);
+
+  return active_ns;
+#endif
+
+#ifdef __APPLE__
+  host_cpu_load_info_data_t cpu_info;
+  mach_msg_type_number_t count = HOST_CPU_LOAD_INFO_COUNT;
+  [[maybe_unused]] const auto ret =
+      host_statistics(mach_host_self(), HOST_CPU_LOAD_INFO, reinterpret_cast<host_info_t>(&cpu_info), &count);
+  DebugAssert(ret == KERN_SUCCESS, "Failed to get host_statistics");
+
+  const auto active_ticks =
+      cpu_info.cpu_ticks[CPU_STATE_SYSTEM] + cpu_info.cpu_ticks[CPU_STATE_USER] + cpu_info.cpu_ticks[CPU_STATE_NICE];
+
+  // The amount of time from HOST_CPU_LOAD_INFO is measured in units of clock ticks.
+  // sysconf(_SC_CLK_TCK) can be used to convert it to ns.
+  const auto active_ns = active_ticks * std::nano::den / (sysconf(_SC_CLK_TCK) * _get_cpu_count());
+
+  return active_ns;
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/*
+ * Returns the time in ns that THIS process has spent on the CPU
+ * since an arbitrary point in the past.
+*/
+uint64_t MetaSystemUtilizationTable::_get_process_cpu_time() {
+  // CLOCK_PROCESS_CPUTIME_ID:
+  // Per-process CPU-time clock (measures CPU time consumed by all threads in the process).

```

Comment:
Not sure what this metric is supposed to mean. How does it differ from user+sys from /proc/self/stat?

Reply:
"If you google around for the patches that went into the kernel to support this (late 2004, by Christoph Lameter), you'll see that it sums the user and system time."

https://arstechnica.com/civis/viewtopic.php?t=1091018

Ok. Maybe add a comment explaining that this is user+sys as known from `time`?

Reply:
The timer_create documentation is better:

>        CLOCK_PROCESS_CPUTIME_ID (since Linux 2.6.12)
              A clock that measures (user and system) CPU time consumed by
              (all of the threads in) the calling process.